src/reconcile/rec_write.c

```diff
@@ -1334,161 +1302,122 @@ __rec_txn_read(WT_SESSION_IMPL *session, WT_RECONCILE *r,
-	 * timestamp.  Also don't recommend lookaside eviction in that case.
-	 */
-	if (__wt_timestamp_cmp(&max_timestamp, &r->stable_timestamp) > 0) {
-		if (!F_ISSET(r, WT_EVICT_UPDATE_RESTORE))
-			return (EBUSY);
-
-		if (!skipped)
-			r->update_mem_uncommitted += update_mem;
-	}
-#endif
-
-	if (F_ISSET(r, WT_EVICT_UPDATE_RESTORE)) {
-		/*
-		 * The save/restore eviction path.
-		 *
-		 * Clear the returned update, it's not needed. If there's an
-		 * on-page key/value pair to which the update list applies, our
-		 * caller writes it to the disk image. If an insert/append list,
-		 * our caller can ignore the key/value pair (everything needed
-		 * is in the update list), or in the case of row-store, write
-		 * the key to the disk image to split up the insert/append list.
-		 */
-		*updp = NULL;
-
-		/* The page can't be marked clean. */
-		r->leave_dirty = true;
-	}
-
-	/*
 	 * The order of the updates on the list matters, we can't move only the
 	 * unresolved updates, move the entire update list.
-	 *
-	 * If we skipped updates, the transaction value is never used.  If we
-	 * didn't skip updates, the list of updates are eventually written to
-	 * the lookaside table, and associated with each update record is the
-	 * transaction ID of the update we wrote in the reconciled page; once
-	 * that transaction ID is globally visible, we know we no longer need
-	 * the lookaside table records, allowing them to be discarded.
 	 */
 	WT_RET(__rec_update_save(session, r, ins, ripcip, *updp));
+	saved = true;
 
 check_original_value:
 	/*
 	 * Returning an update means the original on-page value might be lost,
-	 * and that's a problem if there's a reader that needs it. There are
-	 * two cases: any lookaside table eviction (because the backing disk
-	 * image is rewritten), or any reconciliation of a backing overflow
-	 * record that will be physically removed once it's no longer needed.
+	 * and that's a problem if there's a reader that needs it.  This can
+	 * happen if there is any reconciliation of a backing overflow record
+	 * that will be physically removed once it's no longer needed.
 	 */
-	if (*updp != NULL &&
-	    (F_ISSET(r, WT_EVICT_LOOKASIDE) ||
-	    (vpack != NULL &&
-	    vpack->ovfl && vpack->raw != WT_CELL_VALUE_OVFL_RM)))
-		WT_RET(__rec_append_orig_value(session, page, *updp, vpack));
+	if (F_ISSET(r, WT_REC_LOOKASIDE) || (*updp != NULL && vpack != NULL &&
+	    vpack->ovfl && vpack->raw != WT_CELL_VALUE_OVFL_RM))
+		WT_RET(
+		    __rec_append_orig_value(session, page, first_upd, vpack));
+
+#ifdef HAVE_TIMESTAMPS
+	/* Track the oldest saved timestamp for lookaside. */
+	if (saved && F_ISSET(r, WT_REC_LOOKASIDE)) {
+		for (upd = first_upd; upd->next != NULL; upd = upd->next)
+			;
+		if (__wt_timestamp_cmp(
+		    &r->min_saved_timestamp, &upd->timestamp) > 0)
+			__wt_timestamp_set(
+			    &r->min_saved_timestamp, &upd->timestamp);
+	}
+#else
+	WT_UNUSED(saved);
+#endif
```

Comment:
@michaelcahill: if we append the original on-page value to the update list, we set the append update item's txnid/timestamp to 0. Is it correct to then set `min_saved_timestamp` to 0 as well?